src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_api.py

```diff
@@ -803,3 +805,128 @@ def set_source_mapping(self, py_db, source_filename, mapping):
+
+        def list_children_and_stop_forking(initial_pid, stop=True):
+            children_pids = []
+            if stop:
+                # Ask to stop forking (shouldn't be called for this process, only subprocesses).
+                self._call(
+                    ['kill', '-STOP', str(initial_pid)],
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.PIPE
+                )
+
+            list_popen = self._popen(
+                ['pgrep', '-P', str(initial_pid)],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE
+            )
+
+            if list_popen is not None:
+                stdout, _ = list_popen.communicate()
+                for line in stdout.splitlines():
+                    line = line.decode('ascii').strip()
+                    if line:
+                        pid = str(line)
+                        children_pids.append(pid)
+                        # Recursively get children.
+                        children_pids.extend(list_children_and_stop_forking(pid))
+            return children_pids
+
+        previously_found = set()
+
+        for _ in range(50):  # Try this at most 50 times before giving up.
+
+            children_pids = list_children_and_stop_forking(this_pid, stop=False)
+            found_new = False
+
+            for pid in children_pids:
+                if pid not in previously_found:
+                    found_new = True
+                    previously_found.add(pid)
+                    self._call(
+                        ['kill', '-KILL', str(pid)],
+                        stdout=subprocess.PIPE,
+                        stderr=subprocess.PIPE
+                    )
+
+            if not found_new:
+                break
+
+        # After killing children, kill the current process.
+        os._exit(1)
+
+    def _popen(self, cmdline, **kwargs):
+        try:
+            return subprocess.Popen(cmdline, **kwargs)
+        except:
+            if DebugInfoHolder.DEBUG_TRACE_LEVEL >= 1:
+                pydev_log.exception('Error running: %s' % (' '.join(cmdline)))
+            return None
+
+    def _call(self, cmdline, **kwargs):
+        try:
+            subprocess.check_call(cmdline, **kwargs)
+        except:
+            if DebugInfoHolder.DEBUG_TRACE_LEVEL >= 1:
+                pydev_log.exception('Error running: %s' % (' '.join(cmdline)))
+
+    def set_terminate_child_processes(self, py_db, terminate_child_processes):
+        py_db.terminate_child_processes = terminate_child_processes
+
+    def terminate_process(self, py_db):
+        '''
+        Terminates the current process (and child processes if the option to also terminate
+        child processes is enabled).
+        '''
+        if not py_db.terminate_child_processes:
+            os._exit(1)
+
+        if IS_WINDOWS:
+            self._terminate_process_and_children_windows()
+        else:
+            self._terminate_process_and_children_linux_and_mac()
```

Comment:
My suggestion would be to make these handle the children only (conditional on the flag), and then wrap the call in a `try..finally` that does `os._exit` at the end. This way, if anything goes wrong while processing the subprocess tree, at least the process itself definitely exits. 

The main reason is that the code that deals with children itself spawns another process and then communicates with it, and that can fail in so many ways - e.g. the helper process can spawn successfully but then crash.